Objects/dictobject.c

```diff
@@ -6673,25 +6708,43 @@ _PyObject_StoreInstanceAttribute(PyObject *obj, PyDictValues *values,
         }
 #endif
     }
-    PyDictObject *dict = _PyObject_ManagedDictPointer(obj)->dict;
+
     if (ix == DKIX_EMPTY) {
+        int res;
         if (dict == NULL) {
-            dict = make_dict_from_instance_attributes(
-                    interp, keys, values);
+            dict = materialize_managed_dict_lock_held(obj);
             if (dict == NULL) {
                 return -1;
             }
-            _PyObject_ManagedDictPointer(obj)->dict = (PyDictObject *)dict;
+
+            if (value == NULL) {
+                res = PyDict_DelItem((PyObject *)dict, name);
+            } else {
+                res = PyDict_SetItem((PyObject *)dict, name, value);
+            }
```

Comment:
These calls reacquire the dict's critical section

Reply:
This actually shouldn't.  The reason why the previous comment was there regarding the object not being published yet was because the materialization publishes the dict and therefore it needs to be locked.  I was hoping to avoid re-locking the dict here and publishing it after.  But that complexity didn't quite seem like it was worth it due the extra complexity of the del item case.

But I guess we're doing that in 2 spots already, so I've factored that into a helper and now we'll call that here and publish the materialized dict at the end.